system/lib/wasm_worker/library_wasm_worker.c

```diff
@@ -0,0 +1,240 @@
+
+void __wasm_init_tls(void *memory);
+
+#if !WASM_WORKER_NO_TLS
+__attribute__((constructor(48)))
+static void emscripten_wasm_worker_main_thread_initialize() {
+	uintptr_t* sbrk_ptr = emscripten_get_sbrk_ptr();
+	__wasm_init_tls((void*)*sbrk_ptr);
+	*sbrk_ptr += __builtin_wasm_tls_size();
+}
+#endif
+
+emscripten_wasm_worker_t emscripten_create_wasm_worker_with_tls(void *stackLowestAddress, uint32_t stackSize, void *tlsAddress, uint32_t tlsSize)
+{
+#if WASM_WORKER_NO_TLS
+	return emscripten_create_wasm_worker_no_tls(stackLowestAddress, stackSize);
+#else
+	assert((uintptr_t)stackLowestAddress % 16 == 0);
+	assert(stackSize % 16 == 0);
+	assert(__builtin_wasm_tls_align() != 0 || __builtin_wasm_tls_size() == 0); // Internal consistency check: Clang can report __builtin_wasm_tls_align to be zero if there is no TLS used - double check it never reports zero if it is used.
+	assert(__builtin_wasm_tls_align() == 0 || (uintptr_t)tlsAddress % __builtin_wasm_tls_align() == 0 && "TLS memory address not aligned in a call to emscripten_create_wasm_worker_with_tls()! Please allocate memory with alignment from __builtin_wasm_tls_align() when creating a Wasm Worker!");
+	assert(tlsSize != 0 || __builtin_wasm_tls_size() == 0 && "Program code contains TLS: please use function emscripten_create_wasm_worker_with_tls() to create a Wasm Worker!");
+	assert(tlsSize == __builtin_wasm_tls_size() && "TLS size mismatch! Please reserve exactly __builtin_wasm_tls_size() TLS memory in a call to emscripten_create_wasm_worker_with_tls()");
+	assert(tlsAddress != 0 || tlsSize == 0);
+	return _emscripten_create_wasm_worker_with_tls((void*)stackLowestAddress, stackSize, tlsAddress);
+#endif
+}
+
+emscripten_wasm_worker_t emscripten_create_wasm_worker_no_tls(void *stackLowestAddress, uint32_t stackSize)
+{
+#if !WASM_WORKER_NO_TLS
+	return emscripten_create_wasm_worker_with_tls(stackLowestAddress, stackSize, 0, 0);
+#else
+	assert((uintptr_t)stackLowestAddress % 16 == 0);
+	assert(stackSize % 16 == 0);
+	assert(__builtin_wasm_tls_size() == 0 && "Cannot disable TLS with -sWASM_WORKERS_NO_TLS=1 when compiling code that does require TLS! Rebuild with -sWASM_WORKERS_NO_TLS=1 removed, or remove uses of TLS from the codebase.");
+	return _emscripten_create_wasm_worker_no_tls((void*)stackLowestAddress, stackSize);
+#endif
+}
+
+emscripten_wasm_worker_t emscripten_malloc_wasm_worker(uint32_t stackSize)
+{
+#if WASM_WORKER_NO_TLS
+	return emscripten_create_wasm_worker_no_tls(memalign(16, stackSize), stackSize);
+#else
+	uint32_t tlsSize = __builtin_wasm_tls_size();
+	return emscripten_create_wasm_worker_with_tls(memalign(16, stackSize), stackSize, tlsSize ? memalign(__builtin_wasm_tls_align(), tlsSize) : 0, tlsSize);
+#endif
+}
+
+void emscripten_wasm_worker_sleep(int64_t nsecs)
+{
+	int32_t addr = 0;
+	emscripten_wasm_wait_i32(&addr, 0, nsecs);
+}
+
+void emscripten_lock_init(emscripten_lock_t *lock)
```

Comment:
Instead of creating a whole new set of `emscripten_*` functions as alternatives to the pthread API, what if we wrote and linked in an alternative implementation of the pthread API? That would make it easier to port code to run inside a Wasm Worker.

Reply:
If we wanted to improve the ability to port code to target wasm workers, I think the route to go about that is to implement a new pthread library implementation that backs on top of wasm workers, not one that would replace exposing wasm workers.

Not exposing a direct API here would be a detriment to the simplicity that is gained with exposing direct web primitives, and also to generated code size. When one talks through an existing native Posix API that does not match the web primitives, there is then unfortunately some API inefficiency and cognitive inefficiency that are unavoidable to happen.

The existing Pthread library is already all about the ease of porting. In the past decade we worked a lot to make it behave as 1:1 to native Pthreads as possible. Portability is therefore not a big target for Wasm Workers, but instead directly exposing what web offers is, and the simplicity and minimalness in code size that is gained as a result is.

So it could be an interesting experiment to see what a Wasm Workers based pthreads library would look like, and what would happen e.g. if pthread cancellation points model was removed from its operation altogether. But I think that kind of experiment should not affect how the wasm workers API surface area is publicized, but it is good to stand on its own.